pkg/core/middleware/authn.go

```diff
@@ -0,0 +1,131 @@
+	"net/http"
+	"time"
+
+	"github.com/skygeario/skygear-server/pkg/core/auth"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authinfo"
+	"github.com/skygeario/skygear-server/pkg/core/auth/session"
+	"github.com/skygeario/skygear-server/pkg/core/config"
+	"github.com/skygeario/skygear-server/pkg/core/db"
+	"github.com/skygeario/skygear-server/pkg/core/handler"
+	coreHttp "github.com/skygeario/skygear-server/pkg/core/http"
+	"github.com/skygeario/skygear-server/pkg/core/model"
+	"github.com/skygeario/skygear-server/pkg/core/skydb"
+	"github.com/skygeario/skygear-server/pkg/core/skyerr"
+)
+
+// AuthnMiddleware populate auth context information
+type AuthnMiddleware struct {
+	AuthContextSetter auth.ContextSetter `dependency:"AuthContextSetter"`
+	SessionProvider   session.Provider   `dependency:"SessionProvider"`
+	AuthInfoStore     authinfo.Store     `dependency:"AuthInfoStore"`
+	TxContext         db.TxContext       `dependency:"TxContext"`
+}
+
+// AuthnMiddlewareFactory creates AuthnMiddleware per request.
+type AuthnMiddlewareFactory struct{}
+
+// NewInjectableMiddleware implements InjectableMiddlewareFactory.
+func (f AuthnMiddlewareFactory) NewInjectableMiddleware() InjectableMiddleware {
+	return &AuthnMiddleware{}
+}
+
+// Handle implements InjectableMiddleware.
+func (m *AuthnMiddleware) Handle(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var err error
+		defer func() {
+			if err == nil {
+				next.ServeHTTP(w, r)
+			} else {
+				// clear session cookie if error occurred
+				cookie := &http.Cookie{
+					Name:    coreHttp.CookieNameSession,
+					Path:    "/",
+					Expires: time.Unix(0, 0),
+				}
+				http.SetCookie(w, cookie)
+
+				skyErr := skyerr.NewNotAuthenticatedErr()
+				httpStatus := skyerr.ErrorDefaultStatusCode(skyErr)
+				response := handler.APIResponse{Err: skyErr}
+				encoder := json.NewEncoder(w)
+				w.Header().Set("Content-Type", "application/json")
+				w.WriteHeader(httpStatus)
+				encoder.Encode(response)
+			}
+		}()
+
+		tenantConfig := config.GetTenantConfig(r)
+
+		key := model.CheckAccessKey(tenantConfig, model.GetAPIKey(r))
+		m.AuthContextSetter.SetAccessKey(key)
+
+		accessToken, transport, err := model.GetAccessToken(r)
+		if err != nil {
+			// invalid session token -> must not proceed
+
+			if err == model.ErrTokenConflict {
+				// clear session cookie if session token is conflicted
+				cookie := &http.Cookie{
+					Name:    coreHttp.CookieNameSession,
+					Path:    "/",
+					Expires: time.Unix(0, 0),
+				}
+				http.SetCookie(w, cookie)
+			}
+			return
+		}
+
+		// No access token found. Simply proceed.
+		if accessToken == "" {
+			return
+		}
+
+		if err = m.TxContext.BeginTx(); err != nil {
+			panic(err)
+		}
+		defer m.TxContext.RollbackTx()
+
+		s, err := m.SessionProvider.GetByToken(accessToken, auth.SessionTokenKindAccessToken)
+		if err != nil {
+			// session not found -> treat as no access token is provided
```

Comment:
Example of such case: the access token expires, so session is not found. The client should be notified of this fact and act accordingly. The server should clear the cookie and the client should clear local authn state.

Reply:
Considering calling a public API (i.e. authentication not required), I don't think an expired access token should fail the request?

Reply:
But for endpoint requiring authentication, the client should be notified the session is expired and clean up state. The problem here is that this middleware does not know the whether the underlying handler requires authentication or not.

One solution is to let authn middleware to save the resolution result in request context. The result is `NoSession`, `InvalidSession`, `ValidSession(user)`. Then the authz middleware can make use of this piece information and act accordingly. However, the approach does not work for microservice because we do not have authz for microservice currently. So application must call `/me` on application launch to ensure its authentication state is in sync with the auth gear.